File: find_boxes.py
from PIL import Image
import sys

def find_boxes(file):
    img = Image.open(file).convert("RGBA")
    data = img.load()
    w, h = img.size
    visited = set()
    boxes = []

    for y in range(h):
        for x in range(w):
            if (x, y) in visited: continue
            
            r, g, b, a = data[x, y]
            if a < 128:  # Transparent pixel
                queue = [(x, y)]
                visited.add((x, y))
                min_x, max_x, min_y, max_y = x, x, y, y
                area = 0
                
                while queue:
                    cx, cy = queue.pop()
                    area += 1
                    min_x = min(min_x, cx)
                    max_x = max(max_x, cx)
                    min_y = min(min_y, cy)
                    max_y = max(max_y, cy)
                    
                    for nx, ny in [(cx+1, cy), (cx-1, cy), (cx, cy+1), (cx, cy-1)]:
                        if 0 <= nx < w and 0 <= ny < h:
                            if (nx, ny) not in visited:
                                visited.add((nx, ny))
                                _, _, _, na = data[nx, ny]
                                if na < 128:
                                    queue.append((nx, ny))
                if area > 1000:
                    boxes.append((min_x, min_y, max_x - min_x + 1, max_y - min_y + 1))
            else:
                visited.add((x, y))
                
    boxes.sort(key=lambda b: b[1])
    print(f"--- {file} ---")
    for i, b in enumerate(boxes):
        print(f"Box {i+1}: {{x: {b[0]}, y: {b[1]}, w: {b[2]}, h: {b[3]}}}")

find_boxes("frame4.png")
find_boxes("frame5.png")
find_boxes("frame6.png")

File: find_boxes.js
const Jimp = require('jimp');

async function findBoxes(file) {
  const image = await Jimp.read(file);
  const w = image.bitmap.width;
  const h = image.bitmap.height;
  
  const visited = new Uint8Array(w * h);
  const boxes = [];
  
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      if (visited[y * w + x]) continue;
      
      const idx = (y * w + x) * 4;
      const a = image.bitmap.data[idx + 3];
      
      if (a < 128) {
        // Transparent pixel, start finding bounding box
        let minX = x, maxX = x, minY = y, maxY = y;
        const queue = [{x, y}];
        visited[y * w + x] = 1;
        
        let area = 0;
        
        while (queue.length > 0) {
          const p = queue.pop();
          area++;
          if (p.x < minX) minX = p.x;
          if (p.x > maxX) maxX = p.x;
          if (p.y < minY) minY = p.y;
          if (p.y > maxY) maxY = p.y;
          
          const neighbors = [
            {x: p.x + 1, y: p.y},
            {x: p.x - 1, y: p.y},
            {x: p.x, y: p.y + 1},
            {x: p.x, y: p.y - 1}
          ];
          
          for (const n of neighbors) {
            if (n.x >= 0 && n.x < w && n.y >= 0 && n.y < h) {
              if (!visited[n.y * w + n.x]) {
                const nIdx = (n.y * w + n.x) * 4;
                const nA = image.bitmap.data[nIdx + 3];
                if (nA < 128) {
                  visited[n.y * w + n.x] = 1;
                  queue.push(n);
                }
              }
            }
          }
        }
        
        // Only consider large enough boxes (e.g. > 1000 pixels)
        if (area > 1000) {
          boxes.push({ x: minX, y: minY, w: maxX - minX + 1, h: maxY - minY + 1, area });
        }
      } else {
        visited[y * w + x] = 1;
      }
    }
  }
  
  // Sort boxes by Y
  boxes.sort((a, b) => a.y - b.y);
  console.log(`\n--- ${file} ---`);
  boxes.forEach((b, i) => {
    console.log(`Box ${i+1}: {x: ${b.x}, y: ${b.y}, w: ${b.w}, h: ${b.h}}`);
  });
}

async function run() {
  await findBoxes('frame4.png');
  await findBoxes('frame5.png');
  await findBoxes('frame6.png');
}
run();
